clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -908,311 +628,781 @@ static ParamDesc makeParamDesc(const FieldDecl *Src, QualType Ty) {
+        (void)std::initializer_list<int>{
+            (handlers.handleSyclStreamType(Item, ItemTy), 0)...};
+    }
+  }
+}
+
+// poorly named Parent is the 'how we got here', basically just enough info for
+// the offset adjustment to know what to do about the enter-struct info.
+template <typename ParentTy, typename... Handlers>
+static void VisitAccessorWrapper(CXXRecordDecl *Owner, ParentTy &Parent,
+                                 CXXRecordDecl *Wrapper,
+                                 Handlers &... handlers) {
+  (void)std::initializer_list<int>{(handlers.enterStruct(Owner, Parent), 0)...};
+  VisitAccessorWrapperHelper(Wrapper, Wrapper->bases(), handlers...);
+  VisitAccessorWrapperHelper(Wrapper, Wrapper->fields(), handlers...);
+  (void)std::initializer_list<int>{(handlers.leaveStruct(Owner, Parent), 0)...};
+}
+
+// A visitor function that dispatches to functions as defined in
+// SyclKernelFieldHandler for the purposes of kernel generation.
+template <typename... Handlers>
+static void VisitRecordFields(RecordDecl::field_range Fields,
+                              Handlers &... handlers) {
+#define KF_FOR_EACH(FUNC)                                                      \
+  (void)std::initializer_list<int> { (handlers.FUNC(Field, FieldTy), 0)... }
+
+  for (const auto &Field : Fields) {
+    QualType FieldTy = Field->getType();
+
+    if (Util::isSyclAccessorType(FieldTy))
+      KF_FOR_EACH(handleSyclAccessorType);
+    else if (Util::isSyclSamplerType(FieldTy))
+      KF_FOR_EACH(handleSyclSamplerType);
+    else if (Util::isSyclSpecConstantType(FieldTy))
+      KF_FOR_EACH(handleSyclSpecConstantType);
+    else if (Util::isSyclStreamType(FieldTy)) {
+      // Stream actually wraps accessors, so do recursion
+      CXXRecordDecl *RD = FieldTy->getAsCXXRecordDecl();
+      VisitAccessorWrapper(nullptr, Field, RD, handlers...);
+      KF_FOR_EACH(handleSyclStreamType);
+    } else if (FieldTy->isStructureOrClassType()) {
+      KF_FOR_EACH(handleStructType);
+      CXXRecordDecl *RD = FieldTy->getAsCXXRecordDecl();
+      VisitAccessorWrapper(nullptr, Field, RD, handlers...);
+    } else if (FieldTy->isReferenceType())
+      KF_FOR_EACH(handleReferenceType);
+    else if (FieldTy->isPointerType())
+      KF_FOR_EACH(handlePointerType);
+    else if (FieldTy->isArrayType())
+      KF_FOR_EACH(handleArrayType);
+    else if (FieldTy->isScalarType())
+      KF_FOR_EACH(handleScalarType);
+    else
+      KF_FOR_EACH(handleOtherType);
+  }
+#undef KF_FOR_EACH
+}
+
+// A base type that the SYCL OpenCL Kernel construction task uses to implement
+// individual tasks.
+template <typename Derived> class SyclKernelFieldHandler {
+protected:
+  Sema &SemaRef;
+  SyclKernelFieldHandler(Sema &S) : SemaRef(S) {}
+
+public:
+  // Mark these virutal so that we can use override in the implementer classes,
+  // despite virtual dispatch never being used.
+
+  //// TODO: Can these return 'bool' and we can short-circuit the handling? That
```

Comment:
I think this is a good idea and we should do it.  There is a little more infrastructure involved (since C++14 doesn't have fold expressions ;(), but I think this has benefit.  It allows the field checker to 'reject' a field without the rest of the handlers having to do anything.

IMO, the FieldChecker should probably return isInvalid, that way every other handler will stop receiving fields as soon as we hit an invalid field.

This is a LITTLE more work for the rest of the checkers though, which likely need to return 'false' all the time.

Reply:
This is the one I was talking about, I don't think I can do this without some pretty ugly TMP in a way that short-circuits.  Once we can use Fold Expressions this is a good thing to revisit.

Reply:
So, leave it as is for now?

Reply:
Just remove the TODO.

Reply:
ok.